p2p/src/net/mock/peer.rs

```diff
@@ -74,10 +82,25 @@ where
         p2p_config: Arc<P2pConfig>,
         socket: T::Stream,
         receiver_address: Option<PeerAddress>,
+        handshake_nonce_store: Arc<HandshakeNonceStore>,
         tx: mpsc::Sender<(MockPeerId, PeerEvent)>,
         rx: mpsc::Receiver<MockEvent>,
     ) -> Self {
         let socket = BufferedTranscoder::new(socket);
+
+        // Generate random nonce
+        let handshake_nonce = match role {
+            Role::Inbound => None,
+            Role::Outbound => {
+                let nonce = make_true_rng().gen();
```

Comment:
Two things:

1. This doesn't have to be a true random number. Pseudo-random is more than enough.
2. This number should be generated exactly once when the p2p system is initialized, and never generated again.

I have to say I'm not fully sure how this works together.

Reply:
> This number should be generated exactly once when the p2p system is initialized, and never generated again.

That will be much easier. But are you sure about this? I used random number generator to make fingerprinting more difficult. For example if someday we add TOR addresses then it will be very easy to associate IP address and TOR address of some node. For example Bitcoin Core and Grin use random nonces to detect self-connections.

Reply:
Does bitcoin generate a new number on every connection? If I remember correctly, it's a static variable that they initialize once. Did they change that?

Let's have a call about it.

Reply:
They derive new value for every peer:

```
    NodeId id = GetNewNodeId();
    uint64_t nonce = GetDeterministicRandomizer(RANDOMIZER_ID_LOCALHOSTNONCE).Write(id).Finalize();
```